File: CS/Shared/Common.DataModel/IDataProvider.cs
using DevExpress.Xpf.Data;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;

namespace DevExpress.CRUD.DataModel {
    public interface IDataProvider<T> where T : class {
        IList<T> Read();
        void Create(T obj);
        void Update(T obj);
        void Delete(T obj);
        TResult GetQueryableResult<TResult>(Func<IQueryable<T>, TResult> getResult); //used for virtual sources
        string KeyProperty { get; }
    }
}

File: VB/GridControlCRUD.Common/DataModel - VB/IDataProvider.vb
Imports DevExpress.Xpf.Data
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Linq.Expressions

Namespace DevExpress.CRUD.DataModel
	Public Interface IDataProvider(Of T As Class)
		Function Read() As IList(Of T)
		Sub Create(ByVal obj As T)
		Sub Update(ByVal obj As T)
		Sub Delete(ByVal obj As T)
		Function GetQueryableResult(Of TResult)(ByVal getResult As Func(Of IQueryable(Of T), TResult)) As TResult 'used for virtual sources
		ReadOnly Property KeyProperty() As String
	End Interface
End Namespace
